libcaf_core/caf/flow/op/on_error_resume_next.test.cpp

```diff
@@ -0,0 +1,203 @@
+      THEN("the observer completes with the error") {
+        auto outputs = std::make_shared<std::vector<int>>();
+        auto err = std::make_shared<error>();
+        auto input = std::vector{1, 2, 3, 4};
+        auto completed = std::make_shared<bool>(false);
+        make_observable()
+          .from_container(input)
+          .concat(make_observable().fail<int>(sec::invalid_request))
+          .on_error_resume_next(
+            predicate, make_observable().iota(9).take(2).concat(
+                         make_observable().fail<int>(sec::invalid_request)))
+          .do_on_complete([completed]() { *completed = true; })
+          .do_on_error([err](const error& what) { *err = what; })
+          .for_each([outputs](const int& xs) { outputs->emplace_back(xs); });
+        run_flows();
+        check_eq(*err, error{sec::invalid_request});
+        check_eq(*outputs, std::vector<int>{1, 2, 3, 4, 9, 10});
+        check(!*completed);
+      }
+    }
+  }
+  GIVEN("an observable that fails with sec::runtime_error") {
+    WHEN("calling on_error_resume_next with a predicate that does not match") {
+      THEN("the observer completes with error without using the fallback") {
+        auto outputs = std::make_shared<std::vector<int>>();
+        auto err = std::make_shared<error>();
+        auto input = std::vector{1, 2, 3, 4};
+        auto completed = std::make_shared<bool>(false);
+        make_observable()
+          .from_container(input)
+          .concat(make_observable().fail<int32_t>(sec::runtime_error))
+          .on_error_resume_next(predicate, make_observable().iota(9).take(2))
+          .do_on_complete([completed]() { *completed = true; })
+          .do_on_error([err](const error& what) { *err = what; })
+          .for_each([outputs](const int& xs) { outputs->emplace_back(xs); });
+        run_flows();
+        check_eq(outputs->size(), 4u);
+        check_eq(*err, error{sec::runtime_error});
+        check(!*completed);
+      }
+    }
+  }
+  GIVEN("an observable that completes") {
+    WHEN("calling on_error_resume_next") {
+      THEN("the observer completes with the original observable") {
+        auto outputs = std::make_shared<std::vector<int>>();
+        auto err = std::make_shared<error>();
+        auto completed = std::make_shared<bool>(false);
+        auto input = std::vector{1, 2, 3, 4};
+        make_observable()
+          .from_container(input)
+          .on_error_resume_next(predicate, make_observable().iota(9).take(2))
+          .do_on_error([err](const error& what) { *err = what; })
+          .do_on_complete([completed]() { *completed = true; })
+          .for_each([outputs](const int& xs) { outputs->emplace_back(xs); });
+        run_flows();
+        check_eq(outputs->size(), 4u);
+        check_eq(*err, error{}); // no error
+        check(*completed);
+      }
+    }
+  }
+}
+
+SCENARIO("on_error_resume_next discards items when there is no demand") {
```

Comment:
This test doesn't do what it claims to do. A `multicaster` will buffer items internally if there is no demand. So we will never actually call `on_next` on the operator without demand. To test that, you'll have to cast `uut` back to the actual type and call `on_next` manually.

You could use a `never` as source to make sure there is no interference.